### PR TITLE
IA-2621:  Org unit change request parent shape

### PR DESCRIPTION
**Pull request by beygorghor**
Add alert on whether the change request location falls within the parent's shape or not

Related JIRA tickets : IA-2621
## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- fixed typescript issues for react-leaflet and removed ts-ignore
- added geo_json on parent org unit in api
- hook to see if a location is within a shape
- customise markermap component to get a parent org unit and display it on map
- add popup if marker is not inside his parent shape
- covering edge case when you change parent too

## How to test

- make sure you have an org unit with a parent
- create a change request with this org unit and change the location  with something like this SRID=4326;POINT Z (23.0449 -2.9820 0)
- open change request dialog and check that everything is fine
- if you location is within the parent shape, edit the change request new location value and change lat or long to 0

## Print screen / video

Screenshot 2024-02-01 at 12 51 05

### CODE REVIEW

File: iaso/api/org_unit_change_requests/serializers.py
Comment (quang-le) [inline iaso/api/org_unit_change_requests/serializers.py:1]: Isn't there a python test to update with this change?

File: hat/assets/js/apps/Iaso/domains/app/translations/en.json [hunk 1]
```diff
@@ -717,6 +717,7 @@
     "iaso.map.location": "Location",
     "iaso.map.locationLimit": "Map results limit",
     "iaso.map.locationLimit.error": "Please enter a correct value",
+    "iaso.map.locationNotInShape": "The location is not within the bounds of the shape.",
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/app/translations/en.json]: " the parent's shape"?

File: hat/assets/js/apps/Iaso/components/maps/messages.js [hunk 2]
```diff
@@ -33,6 +33,14 @@ const MESSAGES = defineMessages({
         id: 'iaso.map.title.markerClustering',
         defaultMessage: 'Clustering',
     },
+    parent: {
+        id: 'iaso.label.parent',
+        defaultMessage: 'Parent',
+    },
+    locationNotInShape: {
+        id: 'iaso.map.locationNotInShape',
+        defaultMessage: 'The location is not within the bounds of the shape.',
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/messages.js]: idem: parent shape

File: hat/assets/js/apps/Iaso/components/maps/hooks/useMarkerWithinBounds.ts
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/hooks/useMarkerWithinBounds.ts:1]: ❤️

File: hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx [hunk 3]
```diff
@@ -86,8 +108,8 @@ export const useNewFields = (
             setFields([]);
             return;
         }
-        const orgUnit = changeRequest.org_unit;
-        const fieldDefinitions = {
+        type FieldDefinitions = Record<string, FieldDefinition>;
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx]: This type definition should be moved

File: hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx [hunk 4]
```diff
@@ -139,32 +166,56 @@ export const useNewFields = (
...
+                    false,
+                );
+            }
+
+            return {
+                oldValue,
+                newValue,
+                isChanged,
+            };
+        };
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx]: This function definition should be moved out of the `useMemo` hook

File: hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/reviewChanges/hooks/useNewFields.tsx:1]: The `useMemo`in this hook is used like a `useEffect`. 
`useMemo` is supposed to return a value, so it's confusing that it's not. 

From what I gather, it's used to calculate the value of `fields`, so it should return the value, and `setFields` should be called in a `useEffect` that depends on the return value of `useMemo`. 

That would make the code easier to understand I think

**Review by quang-le — CHANGES_REQUESTED** [review]
Feature works like a charm!
I'm just blocking on the python api test and the useMemo thing

File: iaso/api/org_unit_change_requests/serializers.py
Comment (beygorghor) [reply iaso/api/org_unit_change_requests/serializers.py:1]: you are right, the tests are not using / testing nested org unit type. I'll see if i have time to do it

**Review by quang-le — APPROVED** [review]
